kolibri-flatpak-utils/content_extensions.py

```diff
@@ -0,0 +1,210 @@
+        self.__ref = ref
+        self.__name = name
+        self.__commit = commit
+        self.__content_json = content_json
+
+    @classmethod
+    def from_ref(cls, ref, commit):
+        match = re.match(CONTENT_EXTENSION_RE, ref)
+        if match:
+            name = match.group('name')
+            return cls(ref, name, commit, content_json=None)
+        else:
+            return None
+
+    @classmethod
+    def from_json(cls, json):
+        return cls(
+            json.get('ref'),
+            json.get('name'),
+            json.get('commit'),
+            content_json=json.get('content')
+        )
+
+    def to_json(self):
+        return {
+            'ref': self.ref,
+            'name': self.name,
+            'commit': self.commit,
+            'content': self.content_json
+        }
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
+
+    def __hash__(self):
+        return hash((self.__ref, self.__name, self.__commit))
+
+    @property
+    def ref(self):
+        return self.__ref
+
+    @property
+    def name(self):
+        return self.__name
+
+    @property
+    def commit(self):
+        return self.__commit
+
+    @property
+    def content_json(self):
+        if self.__content_json is not None:
+            return self.__content_json
+
+        content_json_path = os.path.join(self.content_dir, 'content.json')
+
+        try:
+            with open(content_json_path, 'r') as file:
+                self.__content_json = json.load(file)
+        except (OSError, json.JSONDecodeError):
+            self.__content_json = {}
+
+        return self.__content_json
+
+    @property
+    def channels(self):
+        channels_json = self.content_json.get('channels', [])
+        return set(map(KolibriContentChannel.from_json, channels_json))
+
+    @property
+    def base_dir(self):
+        return os.path.join(CONTENT_EXTENSIONS_DIR, self.name)
+
+    @property
+    def content_dir(self):
+        return os.path.join(self.base_dir, 'content')
+
+
+class ContentExtensionsList(object):
+    """
+    Keeps track of a list of content extensions, either cached from a file in
+    $KOLIBIR_HOME, or generated from /.flatpak-info. Multiple lists can be
+    compared to detect changes in the environment.
+    """
+
+    CONTENT_EXTENSIONS_STATE_PATH = os.path.join(KOLIBRI_HOME, 'content-extensions.json')
+
+    def __init__(self, extensions=set()):
+        self.__extensions = set(extensions)
+
+    @classmethod
+    def from_flatpak_info(cls):
+        extensions = set()
+
+        flatpak_info = ConfigParser()
+        flatpak_info.read('/.flatpak-info')
+        app_extensions = flatpak_info.get('Instance', 'app-extensions', fallback='')
+        for extension_str in app_extensions.split(';'):
+            extension_ref, extension_commit = extension_str.split('=')
+            content_extension = ContentExtension.from_ref(extension_ref, commit=extension_commit)
+            if content_extension:
```

Comment:
Would it be worth to call a `content_extension.is_valid()` here to check if the extension has the right format and subdirectories? anyone could create an extension starting with the right ID and it could potentially break Kolibri or the wrapper if we load a broken extension.

Reply:
Good idea :) I'll add that.

Reply:
Okay, done in commit b899f54. Adds a `ContentExtension.is_valid()` method, which confirms the expected directories are in place. Extensions are silently ignored if they don't look like Kolibri content extensions.